lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts

```diff
@@ -61,47 +111,341 @@ export class UnifiedCacheManager {
+        }
+
+        this.setCacheInMemory(cache);
+    }
+
+    /**
+     * Given account key retrieve an account
+     * @param key
+     */
+    getAccount(key: string): AccountEntity {
+        return this.getCacheInMemory().accounts[key] || null;
+    }
+
+    /**
+     * retrieve a credential - accessToken, idToken or refreshToken; given the cache key
+     * @param key
+     */
+    getCredential(key: string): Credential {
+        const cache = this.getCacheInMemory();
+        switch (
+            key.split(Separators.CACHE_KEY_SEPARATOR)[
+                CacheKeyPosition.CREDENTIAL_TYPE
+            ]
+        ) {
+            case "idtoken":
+                return cache.idTokens[key] || null;
+            case "accesstoken":
+                return cache.accessTokens[key] || null;
+            case "refreshtoken":
+                return cache.refreshTokens[key] || null;
+            default:
+                console.log("Cache entity type mismatch");
+                return null;
+        }
+    }
+
+    /**
+     * retrieve accounts matching all provided filters; if no filter is set, get all accounts
+     * not checking for casing as keys are all generated in lower case, remember to convert to lower case if object properties are compared
      * @param homeAccountId
      * @param environment
      * @param realm
      */
-    getAccount(homeAccountId: string, environment: string, realm: string): AccountEntity {
-        const accountCacheKey: Array<string> = [
+    getAccountsFilteredBy(
+        accountFilter: AccountFilter
+    ): AccountCache {
+        return this.getAccountsFilteredByInternal(
+            accountFilter.homeAccountId,
+            accountFilter.environment,
+            accountFilter.realm
+        );
+    }
+
+    /**
+     * retrieve accounts matching all provided filters; if no filter is set, get all accounts
+     * not checking for casing as keys are all generated in lower case, remember to convert to lower case if object properties are compared
+     * @param homeAccountId
+     * @param environment
+     * @param realm
+     */
+    getAccountsFilteredByInternal(
+        homeAccountId?: string,
+        environment?: string,
+        realm?: string
+    ): AccountCache {
+        const accounts: AccountCache = this.getCacheInMemory().accounts;
+        const matchingAccounts: AccountCache = {};
+
+        let matches: boolean = true;
+
+        Object.keys(accounts).forEach((key) => {
+            if (!!homeAccountId) {
+                matches = CacheHelper.matchHomeAccountId(key, homeAccountId);
+            }
+
+            if (!!environment) {
+                matches = matches && CacheHelper.matchEnvironment(key, environment);
+            }
+
+            if (!!realm) {
+                matches = matches && CacheHelper.matchRealm(key, realm);
+            }
+
+            if (matches) {
+                matchingAccounts[key] = accounts[key];
+            }
+        });
```

Comment:
This could be rewritten using `reduce`:

```js
const matchingAccounts: AccountCache = Object.keys(accounts).reduce((matches, key) => {
  const matchHomeAccoutId = !!homeAccountId && CacheHelper.matchHomeAccountId(key, homeAccountId);
  const matchEnvironment = !!environment && CacheHelper.matchEnvironment(key, environment);
  const matchRealm = !!realm && CacheHelper.matchRealm(key, realm);
 
  if (matchHomeAccountId && matchEnvironment && matchRealm) {
    return {
      ...matches,
      [key]: accounts[key]
    }
  }

  return matches;
}, {});

Reply:
@jasonnutter  one question here. The statement if:
```javascript
if (matchHomeAccountId && matchEnvironment && matchRealm) 
```
Does this mandate all three are needed? The intent of the check is to pick any combination of the parameters of the given filter and return what matches. Hence the mutation on `matches` to ensure we skip/add params as passed by the user.

Open to any other way we can get this done.

Reply:
@jasonnutter  This is the only comment not addressed for this PR for now. I will add this once end to end testing is verified, as a part of this issue: #1744